### PR TITLE
feat[contracts]: Use standard RLP transaction format

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Porting over https://github.com/ethereum-optimism/contracts/pull/300/files now that we've converted the predeploys to use `optimistic-solc`. Modifies `OVM_ECDSAContractAccount` and `OVM_SequencerEntrypoint` to only accept standard RLP transactions and not accept our custom format.

### CODE REVIEW

File: l2geth/rollup/sync_service.go [hunk 1]
```diff
@@ -747,7 +745,7 @@ func (s *SyncService) ApplyTransaction(tx *types.Transaction) error {
 	}
 
 	// Set the raw transaction data in the meta
```
Comment (tynes) [inline l2geth/rollup/sync_service.go:747]: Technically the better way to do this is use the raw RLP sent via RPC so that it doesn't need to be reserialized here. The `txmeta` is created in `eth_sendRawTransaction` handler and `rawTransaction` can be set there

**Comment by tynes** [comment]
How do we handle the reverts in the RLP decoding?

File: packages/contracts/contracts/optimistic-ethereum/libraries/wrappers/Lib_ExecutionManagerWrapper.sol [hunk 2]
```diff
@@ -118,6 +118,25 @@ library Lib_ExecutionManagerWrapper {
...
     }
 
+    /**
+     * Calls the ovmCHAINID opcode.
+     * @return Chain ID of the current network.
+     */
+    function ovmCHAINID()
+        internal
+        returns (
+            uint256
```
Comment (karlfloersch) [inline packages/contracts/contracts/optimistic-ethereum/libraries/wrappers/Lib_ExecutionManagerWrapper.sol:128]: @smartcontracts - Did you mention something about using a smaller `uint` for the chainID? Or was that with regard to the nonce that Proto and others have brought up?

I'm not familiar with the DTL RLP codepath but maybe submitting a crazy high `v` would cause unexpected behavior / cause us to decode a transaction incorrectly. Haven't confirmed this but was wondering if it was on your radar

**Comment by smartcontracts** [comment]
> How do we handle the reverts in the RLP decoding?

Reverts in the RLP decoding will cause the whole call frame to revert. So wouldn't get past SequencerEntrypoint.

File: l2geth/rollup/sync_service.go (hunk 1 above)
Comment (smartcontracts) [reply l2geth/rollup/sync_service.go:747]: How would you do this?

File: packages/contracts/test/contracts/OVM/precompiles/OVM_SequencerEntrypoint.spec.ts [hunk 3]
```diff
@@ -94,141 +87,69 @@ describe('OVM_SequencerEntrypoint', () => {
...
-      await expect(
-        Helper_PredeployCaller.callPredeploy(
-          OVM_SequencerEntrypoint.address,
-          calldata
-        )
-      ).to.be.reverted
-    })
+      const call: any = Mock__OVM_ExecutionManager.smocked.ovmCREATEEOA.calls[0]
+      const eoaAddress = ethers.utils.recoverAddress(call._messageHash, {
+        v: call._v + 27,
```
Comment (karlfloersch) [inline packages/contracts/test/contracts/OVM/precompiles/OVM_SequencerEntrypoint.spec.ts:147]: todo(karl): look into why this `v + 27` works when the chain ID of the `DEFAULT_EIP155_TX` is 420

**Review by karlfloersch — COMMENTED** [review]
Awesome! I remember reviewing this a few weeks back & I didn't find anything super note worthy this time. I left a little todo for myself which I'll look into in a bit but just wanted to say that this looks great so far!

Definitely want more eyes but this really simplifies our logic which is great

File: packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol [hunk 4]
```diff
@@ -6,137 +6,56 @@ pragma solidity >0.5.0 <0.8.0;
...
-
-    /*********
-     * Enums *
-     *********/
-    
-    enum TransactionType {
-        NATIVE_ETH_TRANSACTION,
-        ETH_SIGNED_MESSAGE
-    }
+    using Lib_EIP155Tx for Lib_EIP155Tx.EIP155Tx;
```
Comment (ben-chain) [inline packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol:22]: consistency nit: this PR is I think the first time the codebase where we do the `using ... for` syntax.  Not super opinionated that this shouldn't be done but maybe we make a ticket to use this more often if there are other places this could be done.

File: packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol [hunk 5]
```diff
@@ -6,137 +6,56 @@ pragma solidity >0.5.0 <0.8.0;
...
         );
 
-        address target = Lib_ECDSAUtils.recover(
-            encodedTx,
-            isEthSignedMessage,
-            v,
-            r,
-            s
-        );
+        // Value is computed on the fly. Store it to save some gas.
```
Comment (ben-chain) [inline packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol:42]:
```suggestion
        // Value is computed on the fly. Keep it in the stack to save some gas.
```

File: packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol [hunk 6]
```diff
@@ -6,137 +6,56 @@ pragma solidity >0.5.0 <0.8.0;
...
-        )
-    {
-        if (_transactionType == 0) {
-            return TransactionType.NATIVE_ETH_TRANSACTION;
-        } if (_transactionType == 2) {
-            return TransactionType.ETH_SIGNED_MESSAGE;
-        } else {
-            revert("Transaction type must be 0 or 2");
-        }
+        iOVM_ECDSAContractAccount(target).execute(msg.data);
```
Comment (ben-chain) [inline packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol:59]: Since `msg.data` is used up above on line 38, we might get some gas savings if we cache into a `bytes memory` object first so that we are not `CALLDATACOPY`ing to new memory twice.

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol [hunk 7]
```diff
@@ -66,66 +59,46 @@ contract OVM_ECDSAContractAccount is iOVM_ECDSAContractAccount {
...
-        // Need to make sure that the transaction nonce is right and bump it if so.
         require(
-            Lib_ECDSAUtils.recover(
-                _transaction,
-                isEthSign,
-                _v,
-                _r,
-                _s
-            ) == address(this),
+            transaction.sender() == address(this),
```
Comment (ben-chain) [inline packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:72]: Notably the way this code path is currently done is that we re-encode the raw RLP within `sender()` to do recovery.  We could instead recover the sig based on `hash(msg.data[4:end])`.

A quick aside: can you briefly talk about the reasoning for making the `OVM_ECDSAContractAccount` accept encoded `bytes` now as opposed to just a single `Lib_EIP155Tx.EIP155Tx` argument?  Doesn't seem like it makes l2geth any different, and could add redundant encoding/decoding.  Just curious to hear it laid out; makes total sense in the `OVM_SequencerEntryPoint`.

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol [hunk 8]
```diff
@@ -66,66 +59,46 @@ contract OVM_ECDSAContractAccount is iOVM_ECDSAContractAccount {
...
                 msg.sender,
-                SafeMath.mul(decodedTx.gasLimit, decodedTx.gasPrice)
+                SafeMath.mul(transaction.gasLimit, transaction.gasPrice)
             ),
             "Fee was not transferred to relayer."
         );
 
         // Contract creations are signalled by sending a transaction to the zero address.
-        if (decodedTx.to == address(0)) {
+        if (transaction.to == address(0)) {
```
Comment (ben-chain) [inline packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:99]: Shouldn't this be the `isCreate` bool in `Lib_EIP155Tx.EIP155Tx`?

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol [hunk 9]
```diff
@@ -140,7 +113,7 @@ contract OVM_ECDSAContractAccount is iOVM_ECDSAContractAccount {
             // cases, but since this is a contract we'd end up bumping the nonce twice.
             Lib_ExecutionManagerWrapper.ovmINCREMENTNONCE();
 
-            return decodedTx.to.call(decodedTx.data);
+            return transaction.to.call{gas: transaction.gasLimit}(transaction.data);
```
Comment (ben-chain) [inline packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:116]: I think that modifying how gas is handled should probably be treated in a separate PR for separation of concerns, and because it impacts the fee logic other folks are working on.  This syntax [should work though](https://github.com/ethereum-optimism/solidity/blob/bfa0224dd19a3a5101242f281ec81d0f02b9ee3a/libsolidity/codegen/CompilerContext.cpp#L146) once we do want to add it.

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol [hunk 10]
```diff
@@ -66,66 +59,46 @@ contract OVM_ECDSAContractAccount is iOVM_ECDSAContractAccount {
             bytes memory
         )
     {
-        bool isEthSign = _signatureType == Lib_OVMCodec.EOASignatureType.ETH_SIGNED_MESSAGE;
+        // Attempt to decode the transaction.
+        Lib_EIP155Tx.EIP155Tx memory transaction = Lib_EIP155Tx.decode(
+            _encodedTransaction,
+            Lib_ExecutionManagerWrapper.ovmCHAINID()
```
Comment (ben-chain) [inline packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:65]: This can be handled with inline assembly instead of the wrapper if preferable.

File: packages/contracts/contracts/optimistic-ethereum/libraries/wrappers/Lib_ExecutionManagerWrapper.sol [hunk 11]
```diff
@@ -118,6 +118,25 @@ library Lib_ExecutionManagerWrapper {
         return abi.decode(returndata, (address));
     }
 
+    /**
```
Comment (ben-chain) [inline packages/contracts/contracts/optimistic-ethereum/libraries/wrappers/Lib_ExecutionManagerWrapper.sol:121]: This wrapper function is redundant with ability to use `assembly { id := chainid() }` as previously discussed.  If we think that's ugly we can keep it but guessing this was vestigial based on that feedback getting incorporated previously

**Review by ben-chain — COMMENTED** [review]
This is looking great, woot woot!  Seems like Karl has a todo and I had a few small comments, but feels close.

File: l2geth/rollup/sync_service.go (hunk 1 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:747]: This is the line that needs updating: https://github.com/ethereum-optimism/optimism/blob/eef1df4d63fd1b483beda63293ea1746861c3e2a/l2geth/internal/ethapi/api.go#L1738

The RLP encoded tx is in scope as a `hexutil.Bytes`

Comment (tynes) [reply l2geth/rollup/sync_service.go:747]: When the `RollupClient` fetches a tx, it sets the `rawTransaction` here: https://github.com/ethereum-optimism/optimism/blob/eef1df4d63fd1b483beda63293ea1746861c3e2a/l2geth/rollup/client.go#L329

File: packages/contracts/contracts/optimistic-ethereum/libraries/wrappers/Lib_ExecutionManagerWrapper.sol (hunk 2 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/libraries/wrappers/Lib_ExecutionManagerWrapper.sol:128]: Yes, I was planning to make that change in another PR following this one. We'll match the limits up with whatever geth is using. Just didn't want to add that change here b/c too much for a single PR.

File: packages/contracts/test/contracts/OVM/precompiles/OVM_SequencerEntrypoint.spec.ts (hunk 3 above)
Comment (smartcontracts) [reply packages/contracts/test/contracts/OVM/precompiles/OVM_SequencerEntrypoint.spec.ts:147]: What do you mean by this?

File: packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol (hunk 4 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol:22]: We're using it in a few other places but not frequently. I agree that we could consider using this more often.

File: packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol (hunk 6 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol:59]: Interesting, I hadn't thought of that. Do we know that `msg.data` performs a `CALLDATACOPY`?

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol (hunk 7 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:72]: Hmmm yeah interesting, I hadn't thought of passing the `EIP155Tx` directly. I think that's a great idea. Will do this!

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol (hunk 8 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:99]: Ah whoops, forgot to port that over. Thank you!

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol (hunk 9 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:116]: RIP - noticing now that we actually accidentally *removed* the gasLimit logic in this PR from yesterday: https://github.com/ethereum-optimism/optimism/pull/546/files. So might be best to add it back in...

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol (hunk 10 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:65]: I originally had inline assembly but I think the wrapper is preferable.

File: packages/contracts/contracts/optimistic-ethereum/libraries/wrappers/Lib_ExecutionManagerWrapper.sol (hunk 11 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/libraries/wrappers/Lib_ExecutionManagerWrapper.sol:121]: Yeah, I think the assembly is a bit ugly. Wrapper seemed nicer to me.

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol (hunk 7 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:72]: Ok looking at this now, this introduces enough of a diff that we may want to leave it for another PR + treat it as an optimization if this method is too expensive.

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol (hunk 8 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol]: Fixed

File: packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol (hunk 6 above)
Comment (ben-chain) [reply packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol:59]: Yeah, I think it fundamentally has to.  I suppose it could be smart enough to do it once and hold on if it sees duplicates, but I think that's unlikely.

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol (hunk 9 above)
Comment (ben-chain) [reply packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:116]: Ahhhh snap, you are right.  Should've caught that in 546.  In that case let's do it here, but cc @K-Ho who I think has some of the best insight into this right now.

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol (hunk 10 above)
Comment (ben-chain) [reply packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:65]: yeah, I am pretty surprised solidity doesn't expose chainID in a nicer way.

File: packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol (hunk 6 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol]: Just tested this, it does indeed perform two `CALLDATACOPY` operations and it's slightly more efficient to store the value. Just pushed a commit that does this.

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol (hunk 7 above)
Comment (ben-chain) [reply packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:72]: Yeah, it's a bit of a tradeoff.  Ultimately the most gas-efficient version could just require `extcodehash(msg.sender) == SEQUENCER_ENTRYPOINT_CODEHASH` and trust that code path explicitly instead of redoing the work.  Down to leave for future work.

File: packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol (hunk 9 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol:116]: NVM sorry I missed *again* that we were passing `gasleft()` so the first version here was the correct one.
